File: ruby/2_sum/solution.rb
# frozen_string_literal: true

# Algorithm: Two Sum

# Method: Naive
def two_sum(numbers, x)
  (0...numbers.size).each do |i|
    (0...numbers.size).each do |j|
      next if i == j

      return [i, j] if numbers[i] + numbers[j] == x
    end
  end

  return [nil, nil]
end

# Method: 2-pointer
def two_sum_with_2_pointers(numbers, x)
  numbers.sort!

  i = 0
  j = numbers.size  - 1

  while i < j
    sum = numbers[i] + numbers[j]

    if sum == x
      return [i, j]
    elsif sum > x
      j -= 1
    else
      i += 1
    end
  end

  return [nil, nil]
end

# Method: Hash table
def two_sum_with_hash(numbers, x)
  map = {}

  numbers.each_with_index { |number, i| map[x - number] = i }
  numbers.each_with_index { |number, i| return [i, map[number]] if map[number] }

  return [nil, nil]
end

File: go/2_sum/solution.go
// Algorithm: Two Sum
package algo

import (
	"sort"
)

// Method: Naive
func TwoSum(numbers []int, x int) (int, int) {
	for i := 0; i < len(numbers); i++ {
		for j := 0; j < len(numbers); j++ {
			if i == j {
				// next
			} else if numbers[i]+numbers[j] == x {
				return i, j
			}
		}
	}

	return -1, -1
}

// Method: 2-pointer
func TwoSumWith2Pointers(numbers []int, x int) (int, int) {
	sort.Ints(numbers)

	i := 0
	j := len(numbers) - 1

	for i < j {
		sum := numbers[i] + numbers[j]

		if sum == x {
			return i, j
		} else if sum > x {
			j -= 1
		} else {
			i += 1
		}
	}

	return -1, -1
}

// Method: Hash table
func TwoSumWithHash(numbers []int, x int) (int, int) {
	hsh := make(map[int]int)

	for i := 0; i < len(numbers); i++ {
		hsh[x-numbers[i]] = i
	}

	for i := 0; i < len(numbers); i++ {
		value, ok := hsh[numbers[i]]

		if ok {
			return i, value
		}
	}

	return -1, -1
}
